substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,581 @@
+        } catch (Throwable ex) {
+            String description = "InlineDuringParsingAnalysis failed with " + ex.toString();
+            debug.dump(DebugContext.VERBOSE_LEVEL, graph, description);
+            /*
+             * Whatever happens during the analysis is non-fatal because we can just not inline that
+             * invocation.
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        }
+    }
+
+    /**
+     * We collect information during graph construction to filter non-trivial methods and inline
+     * trivial invokes. The result is used in {@link #analyzeMethod}.
+     */
+    class MethodNodeTrackingAndInline extends NodeEventListener implements InlineInvokePlugin {
+
+        final InvocationResultInline result;
+        boolean detectFrameState;
+
+        MethodNodeTrackingAndInline(InvocationResultInline result) {
+            this.result = result;
+            this.detectFrameState = false;
+        }
+
+        @Override
+        public void nodeAdded(Node node) {
+            if (node instanceof ConstantNode) {
+                /* An unlimited amount of constants is allowed. */
+            } else if (node instanceof ParameterNode) {
+                /* Nothing to do, an unlimited amount of parameters is allowed. */
+            } else if (node instanceof ReturnNode) {
+                /*
+                 * Nothing to do, returning a value is fine. We don't allow control flow so there
+                 * can never be more than one return.
+                 */
+            } else if (node instanceof LoadFieldNode) {
+                /* Nothing to do, it's ok to read a static or instance field. */
+            } else if (node instanceof FrameState) {
+                if (!detectFrameState) {
+                    detectFrameState = true;
+                } else {
+                    throw new TrivialMethodDetectorBailoutException("Only framestate for the start node allowed: " + node);
+                }
+            } else if (node instanceof NewArrayNode) {
+                /* Nothing to do, it's ok to have an array of constants */
+            } else {
+                throw new TrivialMethodDetectorBailoutException("Node not allowed: " + node);
+            }
+        }
+
+        @Override
+        public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod callee, ValueNode[] args) {
+            if (callee.getAnnotation(NeverInline.class) != null || callee.getAnnotation(NeverInlineTrivial.class) != null) {
+                return null;
+            }
+
+            CallSite callSite = new CallSite((AnalysisMethod) b.getMethod(), b.bci(), NativeImageInlineDuringParsingPlugin.toAnalysisMethod(callee));
+            InvocationResult state = analyzeMethod(callSite, (AnalysisMethod) callee, args);
+            ImageSingletons.lookup(NativeImageInlineDuringParsingPlugin.class).add(b.getMethod(), b.bci(), (AnalysisMethod) callee, state);
```

Comment:
If this call site is reached by multiple call chains you run the analysis each time, then check if the results match. Wouldn't it be more efficient to run the analysis just once for each call site? What is the performance impact of this analysis on image build time?

Reply:
Yes, you are right. Thanks for the advice. The way of method analysis is changed. We now remember the whole bci chain of calling context.